### PR TITLE
Add support for DrawRangeElements in WebGL2

### PR DESCRIPTION
**Pull request by imiklos**
Adds initial support for the WebGL2 `DrawRangeElements` call.

I have started working on this function, but not sure how could I check for the Uniform Block Backing (https://www.khronos.org/registry/webgl/specs/latest/2.0/#ACTIVE_UNIFORM_BLOCK_BACKING).
I am looking for some advice.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

cc @mmatyas @zakorgy @jdm

### CODE REVIEW

**Comment by jdm** [comment]
This isn't an answer yet, but I believe Firefox's check for active uniform buffer sizes lives in [this code](https://searchfox.org/mozilla-central/source/dom/canvas/WebGLContextDraw.cpp#369-392).

**Comment by jdm** [comment]
We record similar program link info [here](https://github.com/servo/servo/blob/654dd048a3fe4578883ce38616199f06cf850111/components/canvas/webgl_thread.rs#L2118) and store it in WebGLProgram in [here](https://github.com/servo/servo/blob/cf725fc3f59bc1b94e3b35ab4cfe60b3d41559cc/components/script/dom/webglprogram.rs#L157-L191).

**Comment by imiklos** [comment]
@jdm Thank you! I'm working on it. I will update the pull request when I got some progress.

**Comment by imiklos** [comment]
I've just updated the pull request, squashed the commits and updated the test expectations also.

File: components/script/dom/webgl2renderingcontext.rs [hunk 1]
```diff
@@ -191,6 +191,48 @@ impl WebGL2RenderingContext {
         self.base.current_vao_webgl2()
     }
 
+    pub fn validate_uniform_block_for_draw(&self) {
+        if let Some(program) = self.base.current_program() {
```
Comment (jdm) [inline components/script/dom/webgl2renderingcontext.rs]: Let's write this as:
```rust
let program = match self.base.current_program() {
    Some(p) => p,
    None => return,
};
```

File: components/script/dom/webgl2renderingcontext.rs [hunk 2]
```diff
@@ -191,6 +191,48 @@ impl WebGL2RenderingContext {
...
+            program
+                .active_uniform_blocks()
+                .iter()
+                .for_each(|uniform_block| {
```
Comment (jdm) [inline components/script/dom/webgl2renderingcontext.rs]: Let's write this as:
```rust
for uniform_block in &program.active_uniform_blocks() {
    //...
}
```

File: components/script/dom/webgl2renderingcontext.rs [hunk 3]
```diff
@@ -191,6 +191,48 @@ impl WebGL2RenderingContext {
...
+                    let data_size = uniform_block.size as usize;
+                    program.active_uniforms().iter().for_each(|block| {
```
Comment (jdm) [inline components/script/dom/webgl2renderingcontext.rs]: Let's write this as `for block in &program.active_uniforms() {`

File: components/script/dom/webgl2renderingcontext.rs [hunk 4]
```diff
@@ -191,6 +191,48 @@ impl WebGL2RenderingContext {
...
+                        if let Some(index) = block.bind_index {
```
Comment (jdm) [inline components/script/dom/webgl2renderingcontext.rs]: Let's write this as:
```rust
let index = match block.bind_index {
    Some(index) => index,
    None => continue,
};
```

File: components/script/dom/webgl2renderingcontext.rs [hunk 5]
```diff
@@ -191,6 +191,48 @@ impl WebGL2RenderingContext {
...
+                            let indexed = &self.indexed_uniform_buffer_bindings[index as usize];
+                            match indexed.buffer.get() {
```
Comment (jdm) [inline components/script/dom/webgl2renderingcontext.rs]:
```rust
let buffer = match indexed.buffer.get() {
    Some(buffer) => buffer,
    None => {
        self.base.webgl_error(InvalidOperation);
        return;
    },
};
```

File: components/script/dom/webglprogram.rs [hunk 6]
```diff
@@ -574,6 +574,11 @@ impl WebGLProgram {
             return Err(WebGLError::InvalidValue);
         }
 
+        let mut active_uniforms = self.active_uniforms.borrow_mut();
+        if active_uniforms.len() > block_binding as usize {
+            active_uniforms[block_binding as usize].bind_index = Some(block_binding);
```
Comment (jdm) [inline components/script/dom/webglprogram.rs]: Is there any point when we should reset this to None?

**Review by jdm — CHANGES_REQUESTED** [review]
Looking good! Let's try to reduce the rightward-drift in the validation code, though.

File: components/script/dom/webglprogram.rs (hunk 6 above)
Comment (imiklos) [reply components/script/dom/webglprogram.rs]: I'm not sure, but I didn't find any circumstance when we should set this to None except the initial state.

**Comment by imiklos** [comment]
@jdm Thank you for the review! I've addressed what you mentioned and updated the pull request.

**Comment by jdm** [comment]
@bors-servo retry
* #25513

**Comment by jdm** [comment]
@bors-servo retry
* #25513

**Pull request merged**